Adding loopTool to analysis tools

avdudchenko opened the pull request:
## Fixes/Resolves:
This is PR includes the complete loopTool that uses current WaterTAP parametr sweep tool. 

The PR includes the full loopTool and documentation. 
-Support parameter sweeps 
-Support differential parameter sweeps

Currently, the loopTool is not setup to run tests in parallel due to ongoing work of adding a parallel manager, this will need to be updated later on. 


(replace this with the issue # fixed or resolved, if no issue exists then a brief statement of what this PR does)

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

avdudchenko committed:
merged LoopTool from Analysis

avdudchenko committed:
updated diff test

avdudchenko committed:
add_diff_fixed_names

avdudchenko committed:
Update loopTool.rst

avdudchenko committed:
Update index.rst

avdudchenko committed:
fixed tests

avdudchenko committed:
Fixing file pathing, using abspath(__file__)

adam-a-a commented on docs/technical_reference/tools/loopTool.rst:
```diff
@@ -0,0 +1,414 @@
+
+.. currentmodule:: watertap.tools.analysis_tools.loop_tool
+
+The loopTool is a wrapper for the parameter sweep (PS) tool set, and is designed to simplify setting up parametric sweeps, enabling sweeping over discreet design choices, and providing a structured data management.  
```
```suggestion
The loopTool is a wrapper for the parameter sweep (PS) tool set, and is designed to simplify setting up parametric sweeps, enabling sweeping over discrete design choices, and providing structured data management.  
```

adam-a-a commented on docs/technical_reference/tools/loopTool.rst:
```diff
@@ -0,0 +1,414 @@
+
+**Backup management**
+
+The loopTool includes a naïve data management schema to prevent overwriting existing files and minimize simulations runs through creation of backups from existing files. When the loopTool starts it will check if a file with same name already exists, if it does it will rename that file to include a data and time (file_name+_M_D-H_M-S.bak).  
```
```suggestion
The loopTool includes a naïve data management schema to prevent overwriting existing files and minimize simulation runs through creation of backups from existing files. When the loopTool starts it will check if a file with same name already exists, if it does it will rename that file to include a data and time (file_name+_M_D-H_M-S.bak).  
```

k1nshuk committed:
Merge branch 'main' into loopTool

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,677 @@
+        else:
+            return {key: loop_value}
+
+    def setup_multi_processing(self):
```
We should use the parallel manager to setup parallel computing. It has logic for how to handle concurrent futures or MPI depending on the user specification and conda environment.

avdudchenko replied:
This is for the loopTool, as it needs to know if its running with MPI or not for when handling file saving (e.g. getting rank 0), other wise it setups number of subprocess and passes it to parameter sweep tool, which sets up its own parallel management. 
Furthermore, number of subprocesses is only auto defined by looptool, if user does not provide a number them self, as it always opts to run in parallel, this is not the case in parameter sweep tool. We can change this behavior in parametersweep tool (e.g. always revert to multi-processing and auto define number of subprocesses based on available resources). Either way is fine, but the loopTool should not handle the logic for how paramterSweep tool handles multiprocessing, it should just instruct it on what should be done.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_expected_diff_directory.yaml:
```diff
@@ -0,0 +1,64 @@
+          relative_lb: -0.01
+          relative_ub: -0.01
+      diff_samples: 1
+      dir: D:\OneDrive\NAWI_work\Analysis\WaterTAP\watertap-dev\watertap\tools\analysis_tools\loop_tool\tests/output/ro_with_erd/ro_diff_analysis/membrane_cost
```
We should set up temporary paths for testing, ones that are compatible with both windows and unix.

avdudchenko replied:
Those don't need to be tested I guess I should just remove them from the test files. The pathing get tested when loop tool runs.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_expected_sweep_directory.yaml:
```diff
@@ -0,0 +1,76 @@
+            erd_type: pressure_exchanger
+          diff_params: {}
+          diff_samples: 0
+          dir: D:\OneDrive\NAWI_work\Analysis\WaterTAP\watertap-dev\watertap\tools\analysis_tools\loop_tool\tests/output/ro_with_erd/ro_analysis/erd_type/pressure_exchanger/membrane_cost
```
same here - temporary paths

avdudchenko replied:
same as above

avdudchenko committed:
fixed covecod for diff_spec, and pathing in test files

adam-a-a commented on docs/technical_reference/tools/loopTool.rst:
```diff
@@ -0,0 +1,414 @@
+The loopTool requires similar functional setup as PS tool kit. Here we will setup RO_with_energy_recovery.py exmaple flowsheet for use with PS and loopTool.The RO_with_energy_recovery has an option 
+that pass in erd_type allowing user to select type of ERD device (either a pressure_exchanger, pump_as_turbine, or None). The user likely would want to run parameter sweep across these options, and we wil setup the loopTool to do so. 
+
+*The user will need to setup the following three function, description and example are shown below:*
```
```suggestion
*The user will need to setup the following three functions (description and example are shown below):*
```

adam-a-a commented on docs/technical_reference/tools/loopTool.rst:
```diff
@@ -0,0 +1,414 @@
+
+*The user will need to setup the following three function, description and example are shown below:*
+
+1. The *build_function* that builds the flowsheet (ro_build)- This function should build the flowsheet, and accept any kwargs for its configuration. Here we pass explicitly erd_type, but also include *kwargs* incase we want to pass in another options in the future. When loopTool runs, it will pass selectd *kwargs* into this function before running the sweep. 
```
```suggestion
1. The *build_function* that builds the flowsheet (ro_build)- This function should build the flowsheet, and accept any kwargs for its configuration. Here we pass explicitly erd_type, but also include *kwargs* in case we want to pass in other options in the future. When loopTool runs, it will pass selected *kwargs* into this function before running the sweep. 
```

adam-a-a commented on docs/technical_reference/tools/loopTool.rst:
```diff
@@ -0,0 +1,414 @@
+*The user will need to setup the following three function, description and example are shown below:*
+
+1. The *build_function* that builds the flowsheet (ro_build)- This function should build the flowsheet, and accept any kwargs for its configuration. Here we pass explicitly erd_type, but also include *kwargs* incase we want to pass in another options in the future. When loopTool runs, it will pass selectd *kwargs* into this function before running the sweep. 
+2. The *initialization_function* (ro_init)- this function will initialize the model. This is in general optional, as initialization can be done in build function as well, but it is useful for PS tool, as it can use it to reinitialize the model in case a solve fails and try again. 
```
```suggestion
2. The *initialization_function* (ro_init)- this function will initialize the model. This is optional, as initialization can be done in build function as well, but it is useful for the Parameter Sweep tool, as it can use it to reinitialize the model in case a solve fails and try again. 
```

adam-a-a commented on docs/technical_reference/tools/loopTool.rst:
```diff
@@ -0,0 +1,414 @@
+1. The *build_function* that builds the flowsheet (ro_build)- This function should build the flowsheet, and accept any kwargs for its configuration. Here we pass explicitly erd_type, but also include *kwargs* incase we want to pass in another options in the future. When loopTool runs, it will pass selectd *kwargs* into this function before running the sweep. 
+2. The *initialization_function* (ro_init)- this function will initialize the model. This is in general optional, as initialization can be done in build function as well, but it is useful for PS tool, as it can use it to reinitialize the model in case a solve fails and try again. 
+   
+   Additionally, the PS tool (and loopTool) will update the parameters on model tree we are sweep across before calling initialize as well as before calling the solve function. It is critical that in initialize function, we fix or unfix any variable we are sweeping over to their required values for initialize routine unless, that routine can leverage changes in sweeped parameters. 
```
```suggestion
   Additionally, the Parameter Sweep tool (and loopTool) will update the parameters on the model tree that is swept across before calling initialize as well as before calling the solve function. It is critical that in the initialize function, we fix or unfix any variables we are sweeping over to their required values for the initialization routine, unless that routine can leverage changes in sweep parameters. 
```

adam-a-a commented on docs/technical_reference/tools/loopTool.rst:
```diff
@@ -0,0 +1,414 @@
+   
+   Additionally, the PS tool (and loopTool) will update the parameters on model tree we are sweep across before calling initialize as well as before calling the solve function. It is critical that in initialize function, we fix or unfix any variable we are sweeping over to their required values for initialize routine unless, that routine can leverage changes in sweeped parameters. 
+   
+   In our planned sweep, we will be sweeping across flow_mass_phase_comp of NaCl, and RO water recovery. For initialization, these two variables need to be unfixed, so we unfix them at start of ro_init function, as they will be fixed before by PS tool.
```
```suggestion
   In our planned sweep, we will be sweeping across flow_mass_phase_comp of NaCl, and RO water recovery. For initialization. These two variables need to be unfixed, so we unfix them at the start of the ro_init function, as they will have been fixed previously by the Parameter Sweep tool.
```

avdudchenko committed:
switched to useing parallel manager to get mpi config

PR#1102 adds managment for num of subprocess for ray io when running on cluster

avdudchenko committed:
simplify mpi check

avdudchenko committed:
Remove setup_multi_processing

not really needed can be done in place

bknueven committed:
Merge branch 'main' into loopTool

adam-a-a commented on watertap/tools/parameter_sweep/parameter_sweep_reader.py:
```diff
@@ -90,8 +90,70 @@ def get_sweep_params_from_yaml(self, m, yaml_filename):
+                    "relative_lb": values["relative_lb"],
+                    "relative_ub": values["relative_ub"],
+                    "nominal_lb": values["nominal_lb"],
+                    "nominal_ub": values["nominal_ub"],
```
Shouldn't these point to `nominal_lb` and `nominal_ub` based on the logic above regarding `diff_mode` being `precentile` or not?

avdudchenko replied:
Yes fixed that in next commit.

adam-a-a reviewed (commented):
Here are some comments from a first pass. I didn't go through all the details just yet but plan to do a second, more thorough pass.

avdudchenko committed:
Merge branch 'main' into loopTool

avdudchenko committed:
Update docs/technical_reference/tools/loopTool.rst

Co-authored-by: Adam Atia <aatia@keylogic.com>

avdudchenko committed:
Update docs/technical_reference/tools/loopTool.rst

Co-authored-by: Adam Atia <aatia@keylogic.com>

avdudchenko committed:
Update docs/technical_reference/tools/loopTool.rst

Co-authored-by: Adam Atia <aatia@keylogic.com>

avdudchenko committed:
Update docs/technical_reference/tools/loopTool.rst

Co-authored-by: Adam Atia <aatia@keylogic.com>

avdudchenko committed:
Update docs/technical_reference/tools/loopTool.rst

Co-authored-by: Adam Atia <aatia@keylogic.com>

avdudchenko committed:
Update docs/technical_reference/tools/loopTool.rst

Co-authored-by: Adam Atia <aatia@keylogic.com>

avdudchenko committed:
Update docs/technical_reference/tools/loopTool.rst

Co-authored-by: Adam Atia <aatia@keylogic.com>

avdudchenko committed:
update #1 to match new async setup

avdudchenko committed:
updarted tests to reader and small fixes throughout

avdudchenko committed:
added additonal tests

avdudchenko committed:
Update test_data_merging_tool.py

avdudchenko committed:
moved doc to howto as it seems better place for it

avdudchenko committed:
Update index.rst

avdudchenko committed:
Removed legacy code, added additional tests, added MPI test

avdudchenko committed:
Merge branch 'main' into loopTool

avdudchenko committed:
Update mpi4py-test.yml

avdudchenko committed:
mpi fix try #1

avdudchenko committed:
Fixing MPI test

avdudchenko committed:
Fixed MPI rank

avdudchenko committed:
Clean up and changes to MPI rank

avdudchenko committed:
update doc and added outputs and tests

avdudchenko committed:
cleaned up sim_csaes and added key tests for yaml config file

bknueven commented on .github/workflows/mpi4py-test.yml:
```diff
@@ -69,5 +69,12 @@ jobs:
+        # single report
+        coverage combine
+        # convert to XML
+        coverage xml
```
I think we just want to do this step once. But I need to investigate.

You can specify multiple paths with `pytest`, so we should probably just do that instead, e.g.
```
mpiexec -n 2 coverage run --parallel-mode -m mpi4py -m pytest watertap/tools/parameter_sweep/tests/test*parameter_sweep.py watertap/tools/analysis_tools/loop_tool/tests/test*loop_tool.py --no-cov
```

avdudchenko replied:
This would work also. Either way is fine, but I changed it your suggestion, (tests work fine it looks like).

bknueven replied:
Yeah, the tests ran in either case -- but I think only the last test's coverage would be reported correctly. A very minor nit.

avdudchenko committed:
Update mpi4py-test.yml

k1nshuk commented on docs/how_to_guides/how_to_use_loopTool_to_explore_flowsheets.rst:
```diff
@@ -0,0 +1,502 @@
+        optimize_function=ro_setup.ro_solve, # our solve function
+        saving_dir=get_working_dir(), # this gets working directory for script so we can save files in same dirctory
+        save_name="ro_with_erd", # this will be the name of this loop run
+	parallel_backe_end=”RayIo”, #backend we want to use for parallel runs
```
```suggestion
	parallel_back_end=”RayIo”, # backend we want to use for parallel runs
```

k1nshuk commented on docs/how_to_guides/how_to_use_loopTool_to_explore_flowsheets.rst:
```diff
@@ -0,0 +1,502 @@
+            |-solve_successful 
+            |-sweep_params 
+
+We can readely access the data for processing by using h5py. 
```
```suggestion
We can readily access the data for processing by using h5py. 
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+
+from idaes.core.solvers import get_solver
+
+from watertap.tools.parameter_sweep.parameter_sweep import (
```
```suggestion
from watertap.tools.parameter_sweep import (
```
Imports are in __init__.py so it should avoid having to duplicate.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+    RecursiveParameterSweep,
+)
+
+from watertap.tools.parameter_sweep.parameter_sweep_differential import (
```
```suggestion
from watertap.tools.parameter_sweep import (
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+    DifferentialParameterSweep,
+)
+
+from watertap.tools.parameter_sweep.parameter_sweep_reader import ParameterSweepReader
```
```suggestion
from watertap.tools.parameter_sweep import ParameterSweepReader
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+
+from watertap.tools.parameter_sweep.parameter_sweep_reader import ParameterSweepReader
+
+from watertap.tools.parameter_sweep.parameter_sweep_differential import (
+    DifferentialParameterSweep,
+)
```
```suggestion
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+    DifferentialParameterSweep,
+)
+
+# from watertap.tools.analysis_tools.loop_tool.data_merging_tool import *
```
```suggestion
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+import h5py
+import numpy as np
+
+import watertap.tools.MPI as MPI
```
We should get rid of this direct MPI import here or in the next PR.

avdudchenko replied:
This should have been caught in linting and removed in next commit.

k1nshuk committed:
Merge branch 'main' into loopTool

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_data_merging_tool.py:
```diff
@@ -0,0 +1,119 @@
+
+
+def create_test_h5_file():
+    f_name = _this_file_path + "/test_h5_file.h5"
```
We should use the` tmp_path` fixture for writing files in pytest
https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html

avdudchenko replied:
Using tmp_path causes issues as it tries to create same directory multiple times (I get 'directory exists errors). I also want to be able to access the directory in subsequent test, so using a global path works better. This is the same approach taken for testing LSRRO multi-sweep data generation. If this starts causing issues, it can be changed in a different pr.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_expected_sweep_directory.yaml:
```diff
@@ -0,0 +1,218 @@
+            erd_type: pump_as_turbine
+          diff_params: {}
+          diff_samples: 0
+          dir: D:\OneDrive\NAWI_work\Analysis\WaterTAP\watertap-dev\watertap\tools\analysis_tools\loop_tool\tests/output/ro_with_erd/ro_analysis/erd_type/pump_as_turbine/membrane_cost
```
Looks like a hybrid Windows-Unix path. Is this part of the yaml file getting tested in pytest?

avdudchenko replied:
these are auto generated by looptool, but are not actually tested in these files (The test for the data directory is done during the actual file generation process and saving process). These yaml files are meant to check that paramters are properly set up and looping directorties are correctly built.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_expected_sweep_directory.yaml:
```diff
@@ -0,0 +1,218 @@
+            erd_type: pump_as_turbine
+          diff_params: {}
+          diff_samples: 0
+          dir: D:\OneDrive\NAWI_work\Analysis\WaterTAP\watertap-dev\watertap\tools\analysis_tools\loop_tool\tests/output/ro_with_erd/ro_analysis/erd_type/pump_as_turbine/membrane_group
```
Same here

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+
+@pytest.fixture()
+def loop_test_options_setup():
+    cwd = get_working_dir()
```
Use tmp_path

avdudchenko replied:
refer to comment above

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+
+@pytest.fixture()
+def loop_sweep_setup():
+    cwd = get_working_dir()
```
```suggestion
def loop_sweep_setup(tmp_path):
    cwd = tmp_path
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+def loop_diff_setup():
+    # Test without parallel implementation as its broken in
+    # water tap paramtersweep tool
+    cdw = get_working_dir()
```
```suggestion
def loop_diff_setup():
    # Test without parallel implementation as its broken in
    # water tap paramtersweep tool
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+@pytest.mark.component
+def test_failed_setup():
+    with pytest.raises(KeyError, match=r"Unsupported key random_key"):
+        cdw = get_working_dir()
```
```suggestion
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+    with pytest.raises(
+        KeyError, match=r"sweep_param_loop or diff_param_loop not found in config file!"
+    ):
+        cdw = get_working_dir()
```
```suggestion
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+                        test_diff_dict(dicta[key], dictb[key])
+
+                    elif dicta[key] != dictb[key]:
+                        # print(dicta[key], dictb[key])
```
```suggestion
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+        lp, expected_run_dict = loop_sweep_setup
+        lp.build_run_dict()
+
+        def test_diff_dict(dicta, dictb):
```
`test_diff_dict` is defined multiple times in this file.

avdudchenko replied:
fixed

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+    """ used to generate test file"""
+    # with open("test_expected_option_directory.yaml", "w") as file:
+    #     documents = yaml.dump(lp.sweep_directory, file)
+    if has_mpi_peer_processes() == False or (
```
We should not have direct calls for MPI. Can this be abstracted to use the parallel manager?

avdudchenko replied:
This does not do direct MPI call, it uses parallel manager factory.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,370 @@
+    """ used to generate test file"""
+    # with open("test_expected_diff_directory.yaml", "w") as file:
+    #     documents = yaml.dump(lp.sweep_directory, file)
+    if has_mpi_peer_processes() == False or (
```
same here and in the other tests in the file.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+        sweep_samples = 1
+        min_num_samples = None
+        force_rerun = None
+        # print(loop_value, key)
```
```suggestion
```

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+        rank 0 if user running MPI, otherwise do direct
+        solution check"""
+        self.cur_h5_file = (self.h5_file_location, self.h5_directory)
+        if has_mpi_peer_processes():
```
We should use the parallel manager for this.

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,736 @@
+        solution check"""
+        self.cur_h5_file = (self.h5_file_location, self.h5_directory)
+        if has_mpi_peer_processes():
+            mpi_comm = get_mpi_comm_process()
```
Same here, removing direct MPI calls.

avdudchenko committed:
Update docs/how_to_guides/how_to_use_loopTool_to_explore_flowsheets.rst

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
Update watertap/tools/analysis_tools/loop_tool/loop_tool.py

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
Update watertap/tools/analysis_tools/loop_tool/loop_tool.py

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
Update docs/how_to_guides/how_to_use_loopTool_to_explore_flowsheets.rst

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
Update watertap/tools/analysis_tools/loop_tool/loop_tool.py

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
Update watertap/tools/analysis_tools/loop_tool/loop_tool.py

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
Update watertap/tools/analysis_tools/loop_tool/loop_tool.py

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
Update watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py

Co-authored-by: Kinshuk Panda <kinshukpanda@gmail.com>

avdudchenko committed:
updates to loopt tool

avdudchenko committed:
Merge branch 'loopTool' of https://github.com/avdudchenko/watertap into loopTool

avdudchenko committed:
Update test_loop_tool.py

avdudchenko committed:
Update test_loop_tool.py

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/tests/test_loop_tool.py:
```diff
@@ -0,0 +1,332 @@
+    )
+    lp.build_run_dict()
+    """ used to generate test file"""
+    # with open("test_expected_diff_directory.yaml", "w") as file:
+    #     documents = yaml.dump(lp.sweep_directory, file)
```
```suggestion
```

avdudchenko replied:
I want to leave these in, so its easy to recreate the test file in the future. (You can't really create it manually)...

k1nshuk commented on watertap/tools/analysis_tools/loop_tool/loop_tool.py:
```diff
@@ -0,0 +1,732 @@
+            outputs[key] = model.find_component(pyo_object)
+        return outputs
+
+    def run_parameter_sweep(self):
```
Instead of constructing the keyword argument dictionary here and constructing the parameter sweep object, can we call the [parameter_sweep](https://github.com/watertap-org/watertap/blob/971dbce4d86bdff720c3bf6f54823f820a0fa687/watertap/tools/parameter_sweep/parameter_sweep_functions.py#L22) function? Same goes for the differential parameter sweep function later in this file

avdudchenko replied:
I need to use the ps API directly, as some options are not available in parmenter functions (Which IMO should be depreciated, not sure why we even have that). (such as custom_function and custom_function_kwargs).

k1nshuk reviewed (approved):
This is already a big PR so it may be fine to make changes in subsequent PRs.

k1nshuk committed:
Merge branch 'main' into loopTool

k1nshuk committed:
Merge branch 'main' into loopTool

avdudchenko committed:
Update loop_tool.py

lbibl reviewed (approved):
I didn't find notable issues.   Approving at this point. Maybe we count on more testing of them in future's practices.
